clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/PartitionConsumptionState.java

```diff
@@ -582,6 +617,48 @@ public void updateLatestProcessedUpstreamRTOffset(String kafkaUrl, long offset)
     latestProcessedUpstreamRTOffsetMap.put(kafkaUrl, offset);
   }
 
+  public void updateLatestRTOffsetProducedToVTMap(String kafkaUrl, long offset) {
+    latestRTOffsetProducedToVTMap.put(kafkaUrl, offset);
+  }
+
+  public long getLatestRTOffsetProducedToVT(String kafkaUrl) {
+    return latestRTOffsetProducedToVTMap.getOrDefault(kafkaUrl, -1L);
+  }
+
+  public void updateLatestIgnoredUpstreamRTOffset(String kafkaUrl, long offset) {
+    latestIgnoredUpstreamRTOffsetMap.put(kafkaUrl, offset);
+  }
+
+  public long getLatestIgnoredUpstreamRTOffset(String kafkaUrl) {
+    return latestIgnoredUpstreamRTOffsetMap.getOrDefault(kafkaUrl, -1L);
+  }
+
+  public long getLatestProcessedUpstreamRTOffsetWithIgnoredMessages(String kafkaUrl) {
+    long processed = getLatestProcessedUpstreamRTOffset(kafkaUrl);
+    long ignored = getLatestIgnoredUpstreamRTOffset(kafkaUrl);
+    long produced = getLatestRTOffsetProducedToVT(kafkaUrl);
+
+    // we've committed messages at a higher offset then the last thing we ignored. Return the processed offset
+    if (processed >= ignored) {
+      return processed;
+    }
+
+    // We have messages which have been ignored at a higher offset then what we've processed but we still have committed
+    // all messages to local storage that we've produced to upstream VT. In this case, return the ignored offset as the
+    // highest offset. Technically speaking, processed should never be 'greater' then produced, it can at most be
+    // 'equal',
+    // but we'll include the broader comparison as it's still technically correct.
+    if (processed >= produced) {
```

Comment:
As we talked about offline, this causes early RTS announcements.

Reply:
Could you consider renaming `produced` to `downstream` or `toProduce` or  `passed`, or `someGoodName` to reflect the fact that it is updated before producing to Kafka?